feature_engine/encoding/similarity_encoder.py

```diff
@@ -0,0 +1,298 @@
+from typing import List, Optional, Union
+
+import difflib
+import numpy as np
+import pandas as pd
+from sklearn.utils.validation import check_is_fitted
+
+from feature_engine._docstrings.fit_attributes import (
+    _feature_names_in_docstring,
+    _n_features_in_docstring,
+    _variables_attribute_docstring,
+)
+from feature_engine._docstrings.init_parameters import (
+    _ignore_format_docstring,
+    _variables_categorical_docstring,
+)
+from feature_engine._docstrings.methods import _fit_transform_docstring
+from feature_engine._docstrings.substitute import Substitution
+from feature_engine.dataframe_checks import _check_contains_na, check_X
+from feature_engine.encoding.base_encoder import (
+    CategoricalInitMixin,
+    CategoricalMethodsMixin,
+)
+
+
+def _gpm_fast(x1: str, x2: str) -> float:
+    return difflib.SequenceMatcher(None, x1, x2).quick_ratio()
+
+
+_gpm_fast_vec = np.vectorize(_gpm_fast)
+
+
+@Substitution(
+    ignore_format=_ignore_format_docstring,
+    variables=_variables_categorical_docstring,
+    variables_=_variables_attribute_docstring,
+    feature_names_in_=_feature_names_in_docstring,
+    n_features_in_=_n_features_in_docstring,
+    fit_transform=_fit_transform_docstring,
+)
+class StringSimilarityEncoder(CategoricalInitMixin, CategoricalMethodsMixin):
+    """
+    The StringSimilarityEncoder() replaces categorical variables by a set of
+    float variables representing similarity between unique categories in the variable.
+    This new variables will have values in range between 0 and 1, where 0 is the least
+    similar and 1 is the exact match.
+    This encoding is an alternative to OneHotEncoder in the case of poorly
+    defined categorical variables.
+
+    The encoder will create k variables, where k is the number of unique categories.
```

Comment:
I am trying to understand the logic without reading the article.

If a variable was say 10 well defined categories, like Data Analyst, Product Manager etc, you probably want to encode them into 10 float variables, each one would represent 1 category, and the value would be the distance between the value of the observation to the category in question.

If the variable has dirty categories, were the professions are mispelled, instead the user would encode the top categories, hoping that the top ones are spelled correctly. 

Do I get this right?

Reply:
Yes, you are correct. 

On top of that not only misspells/typos, but also categories that are close, like in this example - Data Analyst and Product Analyst should be closer than Data Analyst and Software Engineer. In case of OHE they will be equidistant